File: sdk/rust/examples/provider_workflow/management.rs
use super::{
    config::{field, options, Configuration},
    require, Assertions, ClientProfile, Result, RpcClient,
};
use latent_sdk::management::{
    ApplyPolicyRequest, AuditAckStatus, CapabilityPolicyRecordKind, FailureCategory,
    GetPolicyOperationRequest, GetPolicyRequest, ListCapabilitiesRequest, ListPoliciesRequest,
    ObjectMetadata, OutcomeKnowledge, PageRequest, Policy,
};

fn create_request(
    config: &Configuration,
    operation_id: &str,
    id: &str,
) -> Result<ApplyPolicyRequest> {
    Ok(ApplyPolicyRequest {
        operation_id: operation_id.into(),
        expected_generation: Some(0),
        policy: Some(Policy {
            id: id.into(),
            metadata: Some(ObjectMetadata {
                name: id.into(),
                tenant: Some("tests".into()),
                ..Default::default()
            }),
            document: config.field("policyDocument")?.into(),
            language: "lsf-capability-policy-v1".into(),
            record_kind: CapabilityPolicyRecordKind::POLICY,
            ..Default::default()
        }),
    })
}

pub async fn run(
    config: &Configuration,
    client: &RpcClient,
    assertions: &mut Assertions,
) -> Result<(String, u64)> {
    inspect(config, client).await?;
    assertions.insert("boundedPages", true);
    assertions.insert("providerInspection", true);
    let operation_id = "rust-policy-create".to_owned();
    let id = "rust-example-policy";
    let request = create_request(config, &operation_id, id)?;
    let created = client
        .apply_policy(request.clone(), options())
        .await
        .map_err(|_| "policy-mutation-rpc")?;
    let audit = created
        .metadata
        .audit_ack
        .ok_or("audit-acknowledgement-absent")?;
    require(audit.status == AuditAckStatus::DURABLE, "audit-not-durable")?;
    let attempt = audit
        .attempt_sequence
        .filter(|value| *value > 0)
        .ok_or("audit-attempt-absent")?;
    let receipt = created.value.receipt.ok_or("mutation-receipt-absent")?;
    require(receipt.operation_id == operation_id, "mutation-identity")?;
    let inspected = client
        .get_policy(
            GetPolicyRequest {
                id: id.into(),
                record_kind: CapabilityPolicyRecordKind::POLICY,
            },
            options(),
        )
        .await
        .map_err(|_| "policy-inspection-rpc")?;
    require(
        inspected
            .value
            .policy
            .is_some_and(|value| value.generation == receipt.generation),
        "policy-generation",
    )?;
    let known = client
        .get_policy_operation(
            GetPolicyOperationRequest {
                operation_id: operation_id.clone(),
            },
            options(),
        )
        .await
        .map_err(|_| "policy-operation-rpc")?;
    require(
        known.value.receipt.as_ref() == Some(&receipt),
        "operation-recovery-receipt",
    )?;
    let absent = client
        .get_policy_operation(
            GetPolicyOperationRequest {
                operation_id: "rust-unknown-operation".into(),
            },
            options(),
        )
        .await
        .map_err(|_| "unknown-operation-rpc")?;
    require(
        absent.value.receipt.is_none() && absent.metadata.outcome == OutcomeKnowledge::UNKNOWN,
        "absent-operation-not-unknown",
    )?;
    assertions.insert("mutationReceipt", true);
    let replay = client
        .apply_policy(request.clone(), options())
        .await
        .map_err(|_| "policy-replay-rpc")?;
    require(
        replay.value.receipt.as_ref() == Some(&receipt),
        "exact-replay-receipt",
    )?;
    assertions.insert("exactReplay", true);
    let conflicting = ApplyPolicyRequest {
        operation_id: "rust-stale-precondition".into(),
        ..request
    };
    let failure = client
        .apply_policy(conflicting, options())
        .await
        .err()
        .ok_or("policy-precondition-not-enforced")?;
    require(
        failure.category == FailureCategory::RPC && failure.outcome == OutcomeKnowledge::OBSERVED,
        "policy-precondition-facts",
    )?;
    assertions.insert("preconditionConflict", true);
    Ok((operation_id, attempt))
}

async fn inspect(config: &Configuration, client: &RpcClient) -> Result<()> {
    let request = ListPoliciesRequest {
        record_kind: CapabilityPolicyRecordKind::POLICY,
        page: Some(PageRequest {
            page_size: 1,
            page_token: None,
        }),
    };
    let first = client
        .list_policies(request.clone(), options())
        .await
        .map_err(|_| "policy-first-page-rpc")?;
    require(first.value.policies.len() == 1, "policy-first-page-size")?;
    let token = first
        .value
        .page
        .and_then(|value| value.next_page_token)
        .ok_or("policy-next-token-absent")?;
    let next = client
        .list_policies(
            ListPoliciesRequest {
                page: Some(PageRequest {
                    page_size: 1,
                    page_token: Some(token),
                }),
                ..request
            },
            options(),
        )
        .await
        .map_err(|_| "policy-next-page-rpc")?;
    require(
        next.value.policies.len() == 1 && next.value.policies[0].id != first.value.policies[0].id,
        "policy-page-identity",
    )?;
    let providers = client
        .list_capabilities(
            ListCapabilitiesRequest {
                deployment_id: field(config.target("http"), "route")?.into(),
                page: Some(PageRequest {
                    page_size: 1,
                    page_token: None,
                }),
                ..Default::default()
            },
            options(),
        )
        .await
        .map_err(|_| "capability-inspection-rpc")?;
    require(
        providers.value.capabilities.len() == 1
            && providers.value.capabilities[0].contract == "latent:http/client@0.2.0",
        "provider-inspection-value",
    )
}

File: sdk/go/internal/providerworkflow/management.go
package providerworkflow

import (
	"context"
	"encoding/json"
	"errors"
	"net/url"
	"strconv"

	"latent.dev/sdk/go/profile"
)

func (owner *workflow) management(ctx context.Context) error {
	for _, kind := range []profile.CapabilityPolicyRecordKind{profile.CapabilityPolicyRecordKindPolicy, profile.CapabilityPolicyRecordKindProviderBinding} {
		first, failure := owner.client.ListPolicies(ctx, profile.ListPoliciesRequest{RecordKind: kind, Page: &profile.PageRequest{PageSize: 1}}, profile.CallOptions{})
		if failure != nil || len(first.Value.Policies) != 1 || first.Value.Page == nil || first.Value.Page.NextPageToken == nil || *first.Value.Page.NextPageToken == "" {
			return errors.New("participant-bounded-policy-first-page-failed")
		}
		owner.observe(first.Metadata)
		second, failure := owner.client.ListPolicies(ctx, profile.ListPoliciesRequest{RecordKind: kind,
			Page: &profile.PageRequest{PageSize: 1, PageToken: first.Value.Page.NextPageToken}}, profile.CallOptions{})
		if failure != nil || len(second.Value.Policies) != 1 || second.Value.Policies[0].Id == first.Value.Policies[0].Id ||
			first.Value.CatalogGeneration != second.Value.CatalogGeneration {
			return errors.New("participant-bounded-policy-next-page-failed")
		}
		owner.observe(second.Metadata)
		read, failure := owner.client.GetPolicy(ctx, profile.GetPolicyRequest{Id: first.Value.Policies[0].Id, RecordKind: kind}, profile.CallOptions{})
		if failure != nil || read.Value.Policy == nil || read.Value.Policy.Id != first.Value.Policies[0].Id || read.Value.Policy.RecordKind != kind ||
			read.Value.Policy.Generation != first.Value.Policies[0].Generation || read.Value.Policy.Document == "" {
			return errors.New("participant-policy-provider-record-read-failed")
		}
		owner.observe(read.Metadata)
	}
	owner.result.Assertions["boundedPages"] = true
	for _, name := range []string{"http", "blob"} {
		target := owner.input.Targets[name]
		response, failure := owner.client.ListCapabilities(ctx, profile.ListCapabilitiesRequest{DeploymentId: target.Route, Page: &profile.PageRequest{PageSize: 1}}, profile.CallOptions{})
		if failure != nil || len(response.Value.Capabilities) != 1 || response.Value.Revision == nil {
			return errors.New("participant-provider-inspection-page-failed")
		}
		owner.observe(response.Metadata)
		value := response.Value.Capabilities[0].Inspection
		revision := response.Value.Revision
		if value == nil || value.ProviderBinding == nil || value.ProviderBinding.Id == "" || value.ProviderBinding.Revision == 0 ||
			value.ProviderConfigurationDigest == "" || value.ProviderConfigurationEpoch == 0 || value.ProviderProfile == "" ||
			revision.ComponentDigest != target.ComponentDigest || revision.PublicationId == nil || *revision.PublicationId != target.Publication {
			return errors.New("participant-redacted-provider-identity-missing")
		}
	}
	owner.result.Assertions["providerInspection"] = true
	request := profile.ApplyPolicyRequest{OperationId: owner.result.OperationID, ExpectedGeneration: reference(uint64(0)),
		Policy: &profile.Policy{Id: "go-no-authority", Metadata: &profile.ObjectMetadata{Name: "go-no-authority", Tenant: reference(owner.input.Tenant)},
			RecordKind: profile.CapabilityPolicyRecordKindPolicy, Language: "lsf-capability-policy-v1", Document: owner.input.PolicyDocument}}
	created, failure := owner.client.ApplyPolicy(ctx, request, profile.CallOptions{})
	if failure != nil || created.Value.Receipt == nil || created.Value.Receipt.OperationId != request.OperationId || created.Value.Receipt.Generation == 0 {
		return errors.New("participant-preconditioned-policy-create-failed")
	}
	owner.observe(created.Metadata)
	// lsf-example-begin: management
	lookup, failure := owner.client.GetPolicyOperation(ctx, profile.GetPolicyOperationRequest{OperationId: request.OperationId}, profile.CallOptions{})
	if failure != nil || !sameReceipt(created.Value.Receipt, lookup.Value.Receipt) || lookup.Metadata.Outcome != profile.OutcomeKnowledgeObserved {
		return errors.New("participant-original-policy-operation-recovery-failed")
	}
	owner.observe(lookup.Metadata)
	// lsf-example-end: management
	owner.result.Assertions["mutationReceipt"] = true
	replay, failure := owner.client.ApplyPolicy(ctx, request, profile.CallOptions{})
	if failure != nil || !sameReceipt(created.Value.Receipt, replay.Value.Receipt) {
		return errors.New("participant-exact-explicit-policy-replay-failed")
	}
	owner.observe(replay.Metadata)
	owner.result.Assertions["exactReplay"] = true
	request.Policy.Document = owner.denyDocument()
	_, failure = owner.client.ApplyPolicy(ctx, request, profile.CallOptions{})
	owner.observeFailure(failure)
	if !rpcStatus(failure, 9) && !rpcStatus(failure, 10) {
		return errors.New("participant-changed-document-replay-not-conflicted")
	}
	request.Policy.Document = owner.input.PolicyDocument
	request.OperationId = "go-policy-stale"
	_, failure = owner.client.ApplyPolicy(ctx, request, profile.CallOptions{})
	owner.observeFailure(failure)
	if !rpcStatus(failure, 9) && !rpcStatus(failure, 10) {
		return errors.New("participant-stale-generation-not-conflicted")
	}
	owner.result.Assertions["preconditionConflict"] = true
	return nil
}

func (owner *workflow) denyDocument() string {
	origin, _ := url.Parse(owner.input.UpstreamURL)
	port, _ := strconv.ParseUint(origin.Port(), 10, 16)
	target := owner.input.Targets["http"]
	document := map[string]any{"formatVersion": 1, "tenant": owner.input.Tenant, "rules": []any{map[string]any{
		"id": "go-deny-only", "effect": "deny", "principals": []any{map[string]string{"kind": "user", "subject": "workflow-operator"}},
		"services": []string{target.Service}, "publications": []string{target.Publication}, "capability": "latent:http/client@0.2.0", "operations": []string{"send"},
		"resources": map[string]any{"kind": "http", "origins": []any{map[string]any{"scheme": "http", "host": "localhost", "port": port}},
			"methods": []string{"GET"}, "paths": []string{"/allowed"}, "pathPrefixes": []string{}},
		"ceiling": map[string]uint64{"operations": 0, "inputBytes": 0, "outputBytes": 0, "wallTimeMillis": 0},
	}}}
	encoded, _ := json.Marshal(document)
	return string(encoded)
}
